File: leetcode/medium/TimeBasedKeyValueStore.py
from collections import defaultdict

class TimeMap:
    """
    A time-based key-value store that allows storing multiple values for the same key
    at different timestamps and retrieving the value for a key at a specific timestamp.
    """

    def __init__(self):
        """
        Initializes the TimeMap.
        A dictionary is used to store the key-value pairs, where each key maps to a
        list of (timestamp, value) tuples. The list is kept sorted by timestamp.
        """
        self.store = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        """
        Stores a value for a given key at a specific timestamp.

        Since the problem guarantees that timestamps are strictly increasing for each key,
        we can simply append the new (timestamp, value) tuple to the list for that key.

        Args:
            key: The key to associate the value with.
            value: The value to store.
            timestamp: The timestamp at which to store the value.

        Time complexity: O(1) on average.
        """
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        """
        Retrieves a value for a given key at a specific timestamp.

        It returns the value associated with the largest timestamp that is less than or
        equal to the given timestamp. If no such value exists, it returns an empty string.
        Binary search is used to find the correct value efficiently.

        Args:
            key: The key to retrieve the value for.
            timestamp: The timestamp to retrieve the value at.

        Returns:
            The value for the key at the given timestamp, or an empty string if not found.

        Time complexity: O(log n), where n is the number of values for the key.
        """
        if key not in self.store:
            return ""

        values = self.store[key]
        
        # Binary search to find the value with the largest timestamp <= given timestamp.
        low, high = 0, len(values) - 1
        result = ""

        while low <= high:
            mid = (low + high) // 2
            if values[mid][0] <= timestamp:
                # This is a potential candidate. Store it and look for a better one
                # (i.e., a larger timestamp) in the right half.
                result = values[mid][1]
                low = mid + 1
            else:
                # The timestamp at mid is too high. Look in the left half.
                high = mid - 1
                
        return result

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: leetcode/medium/TimeBasedKeyValueStore.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A time-based key-value store that supports storing multiple values for the same key
 * at different timestamps and retrieving the key's value at a certain timestamp.
 */
class TimeMap {

    // Using a static inner class or record reduces the memory overhead of the outer class reference
    private static record Entry(int timestamp, String value) {}

    private final Map<String, List<Entry>> store;

    public TimeMap() {
        this.store = new HashMap<>();
    }

    /**
     * Stores the key-value pair with the given timestamp.
     * Time Complexity: O(1) average
     */
    public void set(String key, String value, int timestamp) {
        // computeIfAbsent is more concise and slightly more performant than containsKey + get
        this.store.computeIfAbsent(key, k -> new ArrayList<>())
                  .add(new Entry(timestamp, value));
    }

    /**
     * Returns a value such that set was called previously, with timestamp_prev <= timestamp.
     * If there are multiple such values, it returns the one associated with the largest timestamp_prev.
     * If there are no values, it returns "".
     * * Time Complexity: O(log N) where N is the number of values for the specific key.
     */
    public String get(String key, int timestamp) {
        List<Entry> history = store.get(key);
        
        if (history == null || history.isEmpty()) {
            return "";
        }

        // Binary Search for the greatest timestamp <= target timestamp
        int left = 0;
        int right = history.size() - 1;
        String result = "";

        while (left <= right) {
            int mid = left + (right - left) / 2;
            Entry midEntry = history.get(mid);

            if (midEntry.timestamp() <= timestamp) {
                result = midEntry.value();
                left = mid + 1; // Try to find a larger timestamp that still satisfies the condition
            } else {
                right = mid - 1;
            }
        }

        return result;
    }
}